File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def bfs(begin, target, words):
    
    queue = deque([(begin, 0)])
    
    while queue:
        begin, count = queue.popleft()
        
        if begin == target:
            return count
        
        for word in words[:]:
            if wordDiffByOne(begin, word):
                queue.append((word, count +1))
                words.remove(word)
                
def wordDiffByOne(word1, word2):
    return 1 == sum(1 for a, b in zip(word1, word2) if a !=b)

def solution(begin, target, words):
    
    if target not in words:
        return 0
    else:
        return bfs(begin, target, words)

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.Queue;
import java.util.LinkedList;
import java.util.Set;
import java.util.HashSet;

class Record{
    String word;
    int counter;
    
    public Record(String word, int counter){
        this.word = word;
        this.counter = counter;
    }
}

class Solution {
    public int solution(String begin, String target, String[] words) {
        // D.S.
        Queue<Record> q = new LinkedList<>();
        Set<String> visited = new HashSet<>();
        
        // start
        q.add(new Record(begin, 0));
        visited.add(begin);
        int answer = 0;
        
        
        // bfs
        while(!q.isEmpty()){
            Record curr = q.poll();
            String cWord = curr.word;
            int cCounter = curr.counter;
            
            for(int i = 0 ; i<words.length; i++){
                
                if(!visited.contains(words[i]) && isConnected(cWord, words[i])){
                    cCounter++;
                    
                    // end condition => found.
                    if(words[i].equals(target)) return cCounter;
                    
                    visited.add(words[i]);
                    q.add(new Record(words[i], cCounter));
                }
            }
            
        }
       
        return 0; // end condition => not found.
    }
    
    public boolean isConnected(String src, String des){
        int sum = 0;
        for(int i=0; i<src.length(); i++){
           if(src.charAt(i) != des.charAt(i)) sum +=1;
        }
        
        return sum==1;
    }
}